### PR TITLE
Unsub xpath

### PR DESCRIPTION
**Pull request by titusfortner**
This is a great refactor for simplifying the code I've been working on. I also figured out that we were doing some weird things with text field value attribute vs property.

### CODE REVIEW

**Review by p0deje — APPROVED** [review]

File: lib/watir/locators/button/selector_builder/xpath.rb [hunk 1]
```diff
@@ -3,68 +3,57 @@ module Locators
...
-            {xpath: xpath}
-          end
+            @selector.delete(:tag_name)
 
-          protected
+            type = @selector.delete :type
 
-          def use_index?
-            false
+            if type
```
Comment (jkotests) [inline lib/watir/locators/button/selector_builder/xpath.rb]: The type attribute is valid for `button` and `input` elements, so we cannot assume a tag name based on its presence.

For example, given:

~~~~~~~~
<button id="btn_type" type="submit">text</button>
<input id="id_type" type="submit" value="text">
~~~~~~~~

With the new changes:

~~~~~~~~
browser.buttons(type: 'submit').map(&:id)
#=> ["id_type"]
~~~~~~~~

In 6.14:

~~~~~~~~
browser.buttons(type: 'submit').map(&:id)
#=> ["btn_type", "id_type"]
~~~~~~~~

File: lib/watir/locators/button/selector_builder/xpath.rb [hunk 2]
```diff
@@ -3,68 +3,57 @@ module Locators
...
+              "[local-name()='button' or (local-name()='input' and #{input_types(type)})]"
+            else
+              "[local-name()='button']"
+            end
           end
 
           # This is special because text locator for buttons match text or value
-          def add_text
-            return '' unless @selector.key?(:text)
+          def text_string
```
Comment (jkotests) [inline lib/watir/locators/button/selector_builder/xpath.rb:41]: We need to consider the tag name when deciding to check text or value:
* For button elements, `:text` only matches the element text
* For input elements, `:text` only matches the value attribute

Given the following button, which has different text and value attribute:

~~~~~~~~
<button id="btn_dif_value" value="btn_dif_value">btn_dif_text</button>
~~~~~~~~

With these changes, `:text` will now match the text and value:

~~~~~~~~
browser.buttons(text: 'btn_dif_value').map(&:id)
#=> ["btn_dif_value"]
browser.buttons(text: 'btn_dif_text').map(&:id)
#=> ["btn_dif_value"]
~~~~~~~~

Where as 6.14 would only match the text:

~~~~~~~~
browser.buttons(text: 'btn_dif_value').map(&:id)
#=> []
browser.buttons(text: 'btn_dif_text').map(&:id)
#=> ["btn_dif_value"]
~~~~~~~~

**Review by jkotests — DISMISSED** [review]

File: lib/watir/locators/button/selector_builder/xpath.rb (hunk 1 above)
Comment (titusfortner) [reply lib/watir/locators/button/selector_builder/xpath.rb]: grrr, good catch. Thanks.

File: lib/watir/locators/button/selector_builder/xpath.rb (hunk 2 above)
Comment (titusfortner) [reply lib/watir/locators/button/selector_builder/xpath.rb:41]: Ah, if it is an actual button we need to check the text only, and if it is an input we need to check the value only. Got it.

Comment (titusfortner) [reply lib/watir/locators/button/selector_builder/xpath.rb:41]: @jkotests ok, so what is the right answer here:

```html
<button name="button_2" type="submit" value="button_2">Button 2</button>
```

```ruby
expect(browser.button(value: 'Button 2')).to exist
expect(browser.button(value: /Button 2/)).to exist
```

I kind of feel like this should only match `button_2` and not `Button 2`

I guess I should keep it backward compatible for now, though. :-/

Comment (jkotests) [reply lib/watir/locators/button/selector_builder/xpath.rb:41]: Oh. I don't think I ever noticed that. Yeah, I agree, it should be only matching the value attribute. More stuff for the list of things to deprecate.

Comment (titusfortner) [reply lib/watir/locators/button/selector_builder/xpath.rb:41]: I made it backward compatible. I'll add it to the list of things to deprecate. :)

Comment (titusfortner) [reply lib/watir/locators/button/selector_builder/xpath.rb:41]: It's like this logic is complicated or something...

**Pull request merged**